### PR TITLE
MINOR: Catch InvocationTargetException explicitly and propagate underlying cause

### PR DESCRIPTION
**Pull request by divijvaidya**
InvocationTargetException always wraps an underlying cause. It makes sense to catch it as soon as possible and only propagate the underlying cause.

# Change
1. Replace "legacy" [1] with `getCause()`
2. Propagate the cause after catching InvocationTargetException

[1] "legacy as per the docs" https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html

### CODE REVIEW

**Review by Kvicii — APPROVED** [review]
LGTM.

**Comment by divijvaidya** [comment]
@dengziming @showuon please review this small change.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: Have we checked that changing this exception doesn't cause an issue?

File: clients/src/test/java/org/apache/kafka/common/KafkaFutureTest.java
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/KafkaFutureTest.java]: Maybe we should have a helper `invoke` method since we do this many times in this test?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, left a couple of comments.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: Yes. This method is only used in trogdor today and there is no specific handling of ClassNotFoundException exception.

File: clients/src/test/java/org/apache/kafka/common/KafkaFutureTest.java
Comment (divijvaidya) [reply clients/src/test/java/org/apache/kafka/common/KafkaFutureTest.java]: Thanks for your suggestion. I have made the changes that you suggested in the latest commit.

**Comment by divijvaidya** [comment]
@mimaison kindly review when you get a chance.

**Review by mdedetrich — APPROVED** [review]
The changes look good from my end

**Comment by divijvaidya** [comment]
@showuon requesting for your code review time on this one. We already have two approvals from non-committers.

**Comment by divijvaidya** [comment]
The failing tests for JDK 11 are unrelated to the code changes.
```
Build / JDK 11 and Scala 2.13 / testLargeAssignmentAndGroupWithNonEqualSubscription() – org.apache.kafka.clients.consumer.StickyAssignorTest
1m 3s
Build / JDK 11 and Scala 2.13 / testPrepareJoinAndRejoinAfterFailedRebalance() – org.apache.kafka.clients.consumer.internals.EagerConsumerCoordinatorTest
<1s
Build / JDK 11 and Scala 2.13 / testTimeouts() – org.apache.kafka.controller.QuorumControllerTest
```

@ijuma @mimaison requesting your your time to review this. Please note that we have this triaged with two non-committers already.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: Could we throw `ClassNotFoundException` like before, and pass `e.getCause()` to it instead? I think throwing `ClassNotFoundException` for consistency is better.

**Review by showuon — COMMENTED** [review]
@divijvaidya , sorry for the delay. Overall LGTM, left one comment.

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: Wouldn't a `ClassNotFoundException` be wrong in this context and give a wrong impression to user? I am saying this because it's not a problem with Class not being on the JVM's classpath, instead `InvocationTargetException` is thrown when we have found the Class and invoking the constructor is throwing an exception in the line `constructor.newInstance(args`

I am happy to replace KafkaException with other generic RuntimeException that you would suggest but `ClassNotFoundException` appears inappropriate to me.

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: Good point! You're right! When entering here, the constructor is already found and called, just there's exception thrown. Let's keep `KafkaException` like you did. Thank you.

**Review by showuon — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: One downside of this approach is that one cannot catch ReflectiveException to catch all expected exceptions.

Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: True, but I would disagree that it is a downside. 

In my opinion, a caller should be handling the underlying cause because the `InvocationTargetException` is being thrown from business logic inside the constructor. If the caller catches a generic `ReflectiveOperationException` it doesn't get to know (without inspecting details) whether the exception is due to a "reflection op error" such as method not found, class not found etc. or whether it is an exception thrown by the business logic itself. I believe that propagating the underlying business logic exception is better in this scenario.

Having said that, I don't have a strong opinion on this one as it's a minor improvement that will only impact code debuggability when looking at exception traces. Happy to revert if you think otherwise.

Comment (mdedetrich) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: I also agree with @divijvaidya here, having a strong distinction between business logic and reflection based exceptions is ideal and in my opinion over the long term is less brittle (since if you catch business logic exceptions that is unlikely to change)

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: The caller may be itself generic code. For example, see what happened here:

https://github.com/apache/kafka/pull/12230/files#diff-ac079c35d502aee90e8bbfa3b71cd9deda5ad4ffb49364fcc23a9278a848241cR79

The method now throws `Throwable`. It was pretty hard for me to say whether we needed to update some code to handle other exceptions. Is it clear to you?

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: @ijuma , thanks for the explanation. Make sense to me. 
@divijvaidya I think it's better we revert it. WDYT?

Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/utils/Utils.java:466]: @showuon sure, let's do it. I understand the POV where @ijuma is coming from.

**Pull request merged**
